### PR TITLE
first implem of recursive search

### PR DESCRIPTION
**Pull request by kinnou02**

### CODE REVIEW

File: examples/relation.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    match *obj {
+        osmpbfreader::OsmObj::Relation(ref rel) => {
+            if rel.id == 7444 {
+                return true;
+            }else{
+                return false;
+            }
+        }
+        _ => {return false;}
+    }
```
Comment (TeXitoi) [inline examples/relation.rs]: idiomatic:
```rust
    match *obj {
        osmpbfreader::OsmObj::Relation(ref rel) => rel.id == 7444,
        _ => false,
    }
```

File: examples/relation.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use std::collections::HashSet;
+
+fn wanted(obj: &osmpbfreader::OsmObj) -> bool{
+    match *obj {
+        osmpbfreader::OsmObj::Relation(ref rel) => {
+            if rel.id == 7444 {
+                return true;
+            }else{
+                return false;
+            }
```
Comment (TeXitoi) [inline examples/relation.rs]: Really? ;-)

File: examples/relation.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+fn main() {
+    let filename = std::env::args_os().nth(1).unwrap();
+    let path = std::path::Path::new(&filename);
+    let r = std::fs::File::open(&path).unwrap();
+    let mut pbf = osmpbfreader::OsmPbfReader::new(r);
+    let mut finished = false;
+    let mut dependancies = HashSet::new();
+    while !finished{
+        finished = true;
+        for _obj in pbf.iter() {
```
Comment (TeXitoi) [inline examples/relation.rs]: beginning a variable with a `_` means you don't use it. So you should rename `_obj` by `obj`

**Comment by TeXitoi** [comment]
There is also small formating issues, you can use cargo fmt to fix that: https://github.com/rust-lang-nursery/rustfmt

**Comment by TeXitoi** [comment]
Now, just abstract that between a:
```rust
fn get_objs_and_deps<R, F>(reader: &mut OsmPbfReader<R>, pred: F) -> Result<HashMap<OsmId, OsmObj>>
    where R: Read + Seek,
          F: FnMut(&OsmObj) -> bool
{
   unimplemented!()
}
```

See http://texitoi.github.io/osmpbfreader-rs/osmpbfreader/index.html for error handling.

File: examples/relation.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+
+fn main() {
+    let filename = std::env::args_os().nth(1).unwrap();
+    let path = std::path::Path::new(&filename);
+    let r = std::fs::File::open(&path).unwrap();
+    let mut pbf = osmpbfreader::OsmPbfReader::new(r);
+    let mut finished = false;
+    let mut dependancies = HashSet::new();
```
Comment (TeXitoi) [inline examples/relation.rs]: dependencies ?

File: examples/relation.rs (hunk 2 above)
Comment (kinnou02) [inline examples/relation.rs]: yeah! more lines that's always better! :p

File: examples/relation.rs (hunk 1 above)
Comment (TeXitoi) [inline examples/relation.rs]: even better: 
```rust
fn wanted(obj: &osmpbfreader::OsmObj) -> bool {
    obj.id() == osmpbfreader::OsmId::Relation(7444)
}
```

File: src/lib.rs [hunk 5]
```diff
@@ -123,3 +123,46 @@ pub mod groups;
...
+{
+    let mut finished = false;
+    let mut dependencies = HashSet::new();
+    let mut objects = HashMap::new();
+    while !finished{
+        finished = true;
+        for block in reader.primitive_blocks(){
+            let block = try!(block);
+            for obj in blocks::iter(&block) {
+                if dependencies.contains(&obj.id()) || pred(&obj){
```
Comment (TeXitoi) [inline src/lib.rs]: maybe you can `if ! cond { continue; }` to limit indentation.

File: src/lib.rs [hunk 6]
```diff
@@ -123,3 +123,46 @@ pub mod groups;
...
+                    match obj{
+                        OsmObj::Relation(ref rel) => {
+                            for reference in &rel.refs{
+                                if dependencies.insert(reference.member) {
+                                    finished = false;
+                                }
+                            }
```
Comment (TeXitoi) [inline src/lib.rs]: For fun, you can

```rust
if rel.refs.iter().fold(false, |accu, r| deps.insert(*r) || accu) {
    finished = false;
}
```

You can't use `Iterator::any()` as it is lazy.

If you like, you can use it, but if you prefere the iterative version, that's OK for me.

File: src/lib.rs [hunk 7]
```diff
@@ -123,3 +123,46 @@ pub mod groups;
...
+                        }
+                        OsmObj::Way(ref way) => {
+                            for node in &way.nodes {
+                                if dependencies.insert(OsmId::Node(*node)) {
+                                    finished = false;
+                                }
+                            }
+                        }
+                        OsmObj::Node(_) => {}
+                    }
```
Comment (TeXitoi) [inline src/lib.rs]: even more functional:

```rust
finished = match obj {
    OsmObj::Relation(ref rel) => rel.refs.iter().fold(finished, |accu, r| deps.insert(*r) || accu),
    OsmObj::Way(ref way) => way.nodes.iter().fold(finished, |accu, n| deps.insert(OsmId::Node(*n)),
    OsmObj::Node(_) => finished,
};
```

File: src/lib.rs [hunk 8]
```diff
@@ -123,3 +123,46 @@ pub mod groups;
...
+                    objects.insert(obj.id(), obj);
+                }
+            }
+        }
+        try!(reader.rewind());
+    }
+    return Ok(objects);
+
```
Comment (TeXitoi) [inline src/lib.rs]: Empty line

Comment (TeXitoi) [inline src/lib.rs]: And idiomatic way is just ending with `Ok(objects)` (without `;`).

File: examples/relation.rs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+extern crate osmpbfreader;
```
Comment (TeXitoi) [inline examples/relation.rs]: License please.

**Pull request merged**